# Go Programming, Week 5: Overview
* Program Structure 
  * Bitcount
  * Scope
* Basic Data Types
  * Number
  * Boolean
  * String
* Composite Types
  * Array
  * Slice
  * Map
  * Struct
  * JSON and HTML

# Chapter 2. Program Structure

## 2.1 Names

* names of Go functions, variables, constants, types, statement labels, and package
   * begin with a letter (Unicode) or an underscore
   * case sensitive
* Go has 25 keywords (cannot be used as names)
<code>
    break    default     func   interface select
    case     defer       go     map       struct
    chan     else        goto   package   switch
    const    fallthrough if     range     type
    continue for         import return    var
</code>



Predefined names for built-in constants, types, and functions
<code>
Constants:  true false iota nil
[Types](https://golang.org/ref/spec#Types):      int int8 int16 int32 int64
            unit unit8 unit16 uint32 uint64 uintptr
            float32 float64 complex64 complex128
            bool byte rune string error
[Functions](https://golang.org/ref/spec#Built-in_functions):  make len cap new append copy close delete 
            complex real imag 
            panic recover
</code>
* not reserved, and may use in declarations

In [14]:
//package main
import "fmt"

type 整數 int
func 轉換程式(金 整數){
    fmt.Println(金)
} 

func main(){ 
    銅 := 21 
    金 := 3
    轉換程式(11)
    fmt.Println(銅,金)
}

//main()


11
21 3


### Name Visibility and Naming Convention
* Declared within a function
  * Local to the function
* Declared outside of a function
  * Visible in all files of the package 
  * If begin with an uppercase letter, it is exported and visible outside of its package
    * For example, fmt.Printf
* Package names are always in lower case
* No limit on name length
  * The larger the scope of a name, the longer and more meaningful it should be
* Camel case
  * use: QuoteRuneToASCII and parseRequestLine 
  * never: quote_rune_to_ASCII, parse_request_line
  * use: htmlEscape, HTMLEscape, escapeHTML (name case for acronyms and initialisms)
  * never: escapeHtml

## 2.2 Declarations

* Four major kinds of declarations
  * var, type (chap 2)
  * const (chap 3) 
  * func (chap5)
* Go program
  * .go file
  * package declaration (the go file)
  * import declaration
  * package level declarations: types, variables, constants, functions


In [15]:
//gopl.io/ch2/boiling
// Boiling prints the boiling point of water
//package main
import "fmt"

const boilingF = 212.0

func main() {
        var f = boilingF
        var c = (f - 32) * 5 / 9
        fmt.Printf("boiling point = %g°F or %g°C\n", f, c)
        // Output:
        // boiling point = 212°F or 100°C
}

boiling point = 212°F or 100°C


In [29]:
//main()

In [16]:
//gopl.io/ch2/ftoc
import "fmt"

func main() {
	const freezingF, boilingF = 32.0, 212.0
	fmt.Printf("%g°F = %g°C\n", freezingF, fToC(freezingF)) // "32°F = 0°C"
	fmt.Printf("%g°F = %g°C\n", boilingF, fToC(boilingF))   // "212°F = 100°C"
}

func fToC(f float64) float64 {
	return (f - 32) * 5 / 9
}

32°F = 0°C
212°F = 100°C


In [15]:
main()

32°F = 0°C
212°F = 100°C


## 2.3 Variables

* var name type = expression
  <code>
  var s string
  fmt.Println(s)
  var i,j,k int  //int, int, int
  var b,f,s=true, 2.3, "four" //bool, float64, string
  var f, err = os.Open(name) // os.Open returns a file and an error
  </code>

In [6]:
%%
var s string
fmt.Println(s)

In [25]:
import "fmt"
%%
var i,j,k int  //int, int, int
var b,f,s=true, 2.3, "four" //bool, float64, string
fmt.Println(i,j,k,b,f,s)

0 0 0 true 2.3 four


In [11]:
//package main
import "os"

func main(){
    var f, err = os.Open("dup.txt") // os.Open returns a file and an error
    _,_ = f,err
}

* If type is omitted, it is determined by the initializer expression
* If expression is omitted, the initial value is the zero value for that type
  * 0 for numbers
  * false for booleans
  * “” for strings
  * nil for functions and reference types (slice, pointer, map, channel, function)
  * Zero value of a aggregate type (like array and struct): zero value of all elements or fields
* In Go, no uninitialized variables


### 2.3.1 Short Variable Declarations
* Declaration and initialize local variables
  * within functions
  * name := expression
  * type of name is determined by the type of expression
  * A short variable declaration must declare at least one new variable 


In [26]:
import ("image";"image/gif";"math/rand")
%%
nframes := 100
anim := gif.GIF{LoopCount: nframes}
freq := rand.Float64() * 3.0
t := 0.0
fmt.Println(anim,freq,t)

{[] [] 100 [] {<nil> 0 0} 0} 1.8139808639388586 0


* short variable declarations are used to declare and initialize the majority of local variables
* ***var*** declaration for local variable
  * explicit type differs from that of the initializer expression
  * variable will be assigned a value later

In [15]:
%%
i := 100    // an int
var boiling float64 = 100 
// a float64, explicit type differs from that of the initializer 
var names []string
var err error
type Point int
var p Point
fmt.Println(i,boiling, names, err, p)

100 100 [] <nil> 0


In [16]:
%%
i, j := 0, 1
// multiple variables declared and initialized in the same
// short variable declaration
// := is a declaration, = is an assignment
// multiple variable declaration , tuple assignment
i, j = j, i //swap values of i and j

In [17]:
%%
// short variable declarations used for calls to functions
f, err := os.Open("dup.txt")
if err != nil {
    //return err
}
// ... use f ...
f.Close()

In [19]:
%%
//short variable declaration not necessarily declare all the variable
//if some of them already declared, acts like an assignment
in, err := os.Open("dup.txt") // in and err declared
// ...
out, err := os.Create("useless.txt") // out declared, err assigned
fmt.Println(in, err, out)


&{0x14000130180} <nil> &{0x140001301e0}


In [21]:
//nonew.go
//package main
import "os"
func main(){
// short variable declaration must declare at least one new variable
    f, err := os.Open("dup.txt")
//..use of f and err
    f, err := os.Create("useless.txt") //compile error: no new variables
    //to fix
    //f, err = os.Create("useless.txt") //compile error: no new variables
    fmt.Println(f,err)
}

# gonb_5505974d 

 

 Cell[21]: Line 8 
 ./main.go:25:12: no new variables on left side of :=
 
// short variable declaration must declare at least one new variable
 f, err := os.Open("dup.txt")
//..use of f and err
 f, err := os.Create("useless.txt") //compile error: no new variables
 //to fix
 //f, err = os.Create("useless.txt") //compile error: no new variables

ERROR: failed to run "/opt/homebrew/bin/go build -o /var/folders/4d/tg201k8x6yb81f_lrvt4440m0000gn/T/gonb_5505974d/gonb_5505974d": exit status 1

<code>
% go run nonew.go
# command-line-arguments
./nonew.go:4:10: no new variables on left side of :=
</code>

* short variable declaration acts like an assignment only to variables declared in the same lexical block
* declarations in an outer blcok are ignored

### 2.3.2 Pointers

* A pointer value is the address of a variable
* A pointer is the location at which a value is stored

In [23]:
%%
var x int
p := &x // address of x , a pointer to an integer variable, 
   // a value of type *int, pointer to int
   // p points to x, p contains the address of x
   // variable to which p points is *p, the value of the variable
fmt.Println(p)

0x1400009e018


In [24]:
%%
x := 2
p := &x           // p, of type *int, points to x
fmt.Println(*p)   // "1"
*p = 27            // equivalent to x = 2
fmt.Println(x)    // "2"

2
27


In [25]:
%%
var x, y int
fmt.Println(&x==&x, &x==&y, &x== nil) // "true false false"

true false false


In [26]:
var p = f()
func f() *int {
    v := 1
    fmt.Printf("%p\n",&v)
    //print &v value 
    return &v
}
%%
fmt.Println(f() == f()) // "false"

0x14000120018
0x14000120030
0x14000120038
false


In [27]:
func incr(p *int) int {
    *p++ // increments what p points to; does not change p
    return *p
}
%%
v := 1
incr(&v)               // side effect: v is now 2
fmt.Println(incr(&v))  // "3" (and v is 3)

0x14000120018
3


In [19]:
//package main
import (
	"flag"
	"fmt"
	"strings"
)

var n = flag.Bool("n", false, "omit trailing newline")
var sep = flag.String("s", " ", "separator")

func main() {
	flag.Parse()
	fmt.Print(strings.Join(flag.Args(), *sep))
	if !*n {
		fmt.Println()
	}
}


In [20]:
import ("os")
//os.Args = []string{"echo4","-s","/","a","bc","def"};main()
//os.Args = []string{"echo4","-n","a","bc","def"};main()
//os.Args = []string{"echo4","-help"}
//main()
//%args -s / a bc def
//%args -help
%args -n a bc def
func main() {
	flag.Parse()
	fmt.Print(strings.Join(flag.Args(), *sep))
	if !*n {
		fmt.Println()
	}
}

a bc def

<code>
$ go install gopl.io/ch2/echo4@latest

$ bin/echo4 -help
Usage of ./echo4:

$ bin/echo -s / a bc def
a/bc/def

$ bin/echo -n a bc def
a bc def$
</code>

### 2.3.3 The new Function

In [2]:
import "fmt"
//x := 1
//p := &x
%%
p := new(int) //p, of type *int, points to an unnamed int variable
fmt.Println(*p) // "0"
*p = 2 
fmt.Println(*p) // "2"

0
2


In [3]:
//two newInt functions have identical behaviors
func newInt() *int {
    return new(int)
}

In [4]:
func newInt() *int {
    var dummy int
    return &dummy
}

In [6]:
//each call to new returns a distinct variable with unique address
%%
p := new(int)
q := new(int)
fmt.Println(p == q) // "false"

false


In [7]:
// type carries no information with size zero
%%
a := new(struct{})
b := new([0]int)
fmt.Printf("%p, %p\n",a,b) // the same address value

0x1028172d8, 0x1028172d8


### 2.3.4 Lifetime of Variables

<code>
for t:= 0.0; t < cycles * 2 *math.Pi; i+= res {
    x := math.Sin(t)
    y := math.Sin(t*freq+ phase)
    img.SetColorIndex( size + int(x*size+0.5), size + int(y*size + 0.5), blackIndex)
}
</code>
the variable t is created each time the for loop begins, and variables x and y are created on each iteration of the loop

In [8]:
var global *int
func f() {
    var x int       //x should be heap allocated
    x = 3
    global = &x
} //x escape from f

func g() {
    y := new(int)  //y can be on the stack
    *y = 1
}//y doesn't escape from g, it can be on the stack

//fmt.Println(*global)
%%
f()
fmt.Println(*global)

3


## 2.4. Assignments

 <code>
    x = 1                      // named variable
    *p = true                  // indirect variable
    person.name = "bob"        // struct field
    count[x] = count[x] * scale // array or slice or map element
    // count[x] *= scale
 </code>

In [9]:
%%
v := 1
v++ // same as v = v + 1; v becomes 2
v-- // same as v = v - 1; v becomes 1 again
fmt.Println(v)

1


### 2.4.1. Tuple Assignment

In [10]:
%%
a :=[]int {9,6,4,10,11,12,13}
i, j := 5, 2
x, y := 3, 9

x , y = y, x
a[i], a[j] = a[j], a[i]     //swapping the values
fmt.Println(a)    
fmt.Println(x,y)

[9 6 12 10 11 4 13]
9 3


In [11]:
func gcd(x,y int) int {
    for y != 0 {
        x, y = y, x%y
    }
    return x
}
func fib(n int) int {
    x, y := 0, 1
    for i := 0; i< n; i++ {
        x, y = y, x+y
    }
    return x
}

In [12]:
%%
fmt.Println(gcd(80,8), gcd(24, 120), gcd(37, 39))
fmt.Println(fib(1), fib(2), fib(3), fib(4), fib(5), fib(6), fib(7), fib(8), fib(9), fib(10))

8 24 1
1 1 2 3 5 8 13 21 34 55


<code>
i, j, k = 2, 3, 5 //tuple assignment
f, err = os.Open("dup.txt") // function calls returns two values
v, ok = m[key]   // map lookup
v, ok = x.(T)    // type assertion, 
    //if interface value of x is of type t, 
    //v will be the value of x, and ok is true. 
    //Otherwise, v is zero, and ok is false
v, ok = <-ch     // channel receive
</code>

<code>
_, err = io.Copy(dst, src) // discard byte count
_, ok  = x.(T)    //check type but discard result
</code>

### 2.4.2 Assignability

* Assignment statements are an explicit form of assignment
* implicit assignment
  * function call assigns argument values to parameter variables
  * return statement implicitly assigns return operands to the return variables
  * literal expression for a composite type (in the following example)

In [13]:
%%
medals := []string{"gold","silver","brone"}
fmt.Println(medals)
//implicitly assigns each elements, like the following
//elements of maps and channels are similar 

[gold silver brone]


In [15]:
%%
var medals [3]string
medals[0] = "gold"
medals[1] = "silver"
medals[2] = "brone"
fmt.Println(medals)

[gold silver brone]


### Assignability
* types must exactly match
* nil can be assigned to any variable of interface or reference type
* as of ==, != , the first operand must be assignable to type of the second operand

## 2.5. Type Declarations

* type of a variable
  * defines the characteristics of the value it may take on
  * size (number of bits or number of elements)
  * how they are represented
  * operations can be performed on them
  * methods associated with them
* A *type* declaration defines a new named type
* ***type name underlying-type***
  * appear at package level visible throughout the package
  * if exported (upper-case letter), accessible from other package

* For every T, there is a corresponding conversion operation T(x)
  * Convert the value to type T
  * Conversion from one type to another is allowed
    * both have the same underlying type
    * numeric types
      * a float to an integer, discard the fractional part
    * string and slice types
      * string to []byte slice, allocates a copy of the string

In [16]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

// See page 39.
//!+
//gopl.io/ch2/tempconv0/celsius.go
// Package tempconv performs Celsius and Fahrenheit temperature computations.
//package tempconv

import "fmt"

type Celsius float64
type Fahrenheit float64

const (
	AbsoluteZeroC Celsius = -273.15
	FreezingC     Celsius = 0
	BoilingC      Celsius = 100
)

func CToF(c Celsius) Fahrenheit { return Fahrenheit(c*9/5 + 32) }

func FToC(f Fahrenheit) Celsius { return Celsius((f - 32) * 5 / 9) }

//!-

func (c Celsius) String() string { return fmt.Sprintf("%g°C", c) }
func (f Fahrenheit) String() string { return fmt.Sprintf("%g°F", f) }

In [18]:
%%
var c Celsius
var f Fahrenheit
fmt.Println(c == 0)
fmt.Println(f >= 0)

fmt.Println(c == Celsius(f))

true
true
true


In [19]:
%%
var c Celsius
var f Fahrenheit
fmt.Println(c == f)

# gonb_a832fa65 

 

 Cell[19]: Line 4 
 ./main.go:65:19: invalid operation: c == f (mismatched types Celsius and Fahrenheit)
 
	flag.Parse()
	var c Celsius
	var f Fahrenheit
 fmt.Println(c == f)
 
}

ERROR: failed to run "/opt/homebrew/bin/go build -o /var/folders/4d/tg201k8x6yb81f_lrvt4440m0000gn/T/gonb_a832fa65/gonb_a832fa65": exit status 1

In [22]:
%%
fmt.Printf("%g\n", BoilingC-FreezingC) // "100" °C
boilingF := CToF(BoilingC)
fmt.Printf("%g\n", boilingF-CToF(FreezingC)) // "180" °F

100
180


In [4]:
%%
c := FToC(212.0)
fmt.Println(c.String()) // "100°C"
fmt.Printf("%v\n", c)   // "100°C"; no need to call String explicitly
fmt.Printf("%s\n", c)  // "100°C"
fmt.Println(c)          // "100°C"
fmt.Printf("%g\n", c)   // "100"; does not call String
fmt.Println(float64(c)) // "100"; does not call String

100°C
100°C
100°C
100°C
100
100


In [24]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

//package tempconv

import "fmt"

func Example_one() {
	{
		//!+arith
		fmt.Printf("%g\n", BoilingC-FreezingC) // "100" °C
		boilingF := CToF(BoilingC)
		fmt.Printf("%g\n", boilingF-CToF(FreezingC)) // "180" °F
		//!-arith
	}
	/*
		//!+arith
		fmt.Printf("%g\n", boilingF-FreezingC)       // compile error: type mismatch
		//!-arith
	*/

	// Output:
	// 100
	// 180
}

func Example_two() {
	//!+printf
	c := FToC(212.0)
	fmt.Println(c.String()) // "100°C"
    fmt.Printf("%v\n", c)   // "100°C"; no need to call String explicitly
    fmt.Printf("%s\n", c)  // "100°C"
	fmt.Println(c)          // "100°C"
	fmt.Printf("%g\n", c)   // "100"; does not call String
	fmt.Println(float64(c)) // "100"; does not call String
	//!-printf

	// Output:
	// 100°C
	// 100°C
	// 100°C
	// 100°C
	// 100
	// 100
}

In [21]:

!go test gopl.io/ch2/tempconv0

ok  	gopl.io/ch2/tempconv0	(cached)


<code>
$go test gopl.io/ch2/tempconv0
ok      gopl.io/ch2/tempconv0   0.020s
</code>

## 2.6. Packages and Files

In [34]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

// See page 43.
//!+

// Cf converts its numeric argument to Celsius and Fahrenheit.
//package main

import (
	"fmt"
	"os"
	"strconv"

	"gopl.io/ch2/tempconv"
)
%args 32
func main() {
	for _, arg := range os.Args[1:] {
		t, err := strconv.ParseFloat(arg, 64)
		if err != nil {
			fmt.Fprintf(os.Stderr, "cf: %v\n", err)
			os.Exit(1)
		}
		f := tempconv.Fahrenheit(t)
		c := tempconv.Celsius(t)
		fmt.Printf("%s = %s, %s = %s\n",
			f, tempconv.FToC(f), c, tempconv.CToF(c))
	}
}

//!-

32°F = 0°C, 32°C = 89.6°F


In [1]:
import ("os")
os.Args = []string{"cf","32"} 
main()
//main()

ERROR: repl.go:3:1: undefined identifier: main

In [41]:
//!go install gopl.io/ch2/cf@latest
!./ch2/cf/cf 32
!./ch2/cf/cf 212
!./ch2/cf/cf -40

32°F = 0°C, 32°C = 89.6°F
212°F = 100°C, 212°C = 413.6°F
-40°F = -40°C, -40°C = -40°F


<code>
$ go install gopl.io/ch2/cf@latest
$./cf 32
32°F = 0°C, 32°C = 89.6°F
$ ./cf 212
212°F = 100°C, 212°C = 413.6°F
$ ./cf -40
-40°F = -40°C, -40°C = -40°F
</code>

### 2.6.2. Package Initialization

In [42]:
import "fmt"
var a = b + c    // a initialized third, to 3
var b = f()      // b initialized second, to 2, by callin f
var c = 1        // c initialized first, to 1
func f() int {return c+1}
%%
fmt.Println(a,b,c)

3 2 1


In [27]:
func PopCount(x uint64) int {
	return int(pc[byte(x>>(0*8))] +
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))])
}
var pc [256]byte
%%
for i := range pc {
	pc[i] = pc[i/2] + byte(i&1)
}
fmt.Printf("%v\n",pc)
x := uint64(0xffed19)
fmt.Printf("%b:%d\n",x,PopCount(x))

[0 1 1 2 1 2 2 3 1 2 2 3 2 3 3 4 1 2 2 3 2 3 3 4 2 3 3 4 3 4 4 5 1 2 2 3 2 3 3 4 2 3 3 4 3 4 4 5 2 3 3 4 3 4 4 5 3 4 4 5 4 5 5 6 1 2 2 3 2 3 3 4 2 3 3 4 3 4 4 5 2 3 3 4 3 4 4 5 3 4 4 5 4 5 5 6 2 3 3 4 3 4 4 5 3 4 4 5 4 5 5 6 3 4 4 5 4 5 5 6 4 5 5 6 5 6 6 7 1 2 2 3 2 3 3 4 2 3 3 4 3 4 4 5 2 3 3 4 3 4 4 5 3 4 4 5 4 5 5 6 2 3 3 4 3 4 4 5 3 4 4 5 4 5 5 6 3 4 4 5 4 5 5 6 4 5 5 6 5 6 6 7 2 3 3 4 3 4 4 5 3 4 4 5 4 5 5 6 3 4 4 5 4 5 5 6 4 5 5 6 5 6 6 7 3 4 4 5 4 5 5 6 4 5 5 6 5 6 6 7 4 5 5 6 5 6 6 7 5 6 6 7 6 7 7 8]
111111111110110100011001:17


In [45]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

// See page 45.

// (Package doc comment intentionally malformed to demonstrate golint.)
//!+
//package popcount

// pc[i] is the population count of i.
var pc [256]byte

func init() {
	for i := range pc {
		pc[i] = pc[i/2] + byte(i&1)
	}
}
// 0:0 , 1:1, 10:1, 11:2, 100:1,101:2,110:2 111:3, 1000:1, 

// PopCount returns the population count (number of set bits) of x.
func PopCount(x uint64) int {
	return int(pc[byte(x>>(0*8))] +
		pc[byte(x>>(1*8))] +
		pc[byte(x>>(2*8))] +
		pc[byte(x>>(3*8))] +
		pc[byte(x>>(4*8))] +
		pc[byte(x>>(5*8))] +
		pc[byte(x>>(6*8))] +
		pc[byte(x>>(7*8))])
}

//!-

In [2]:
// Copyright © 2016 Alan A. A. Donovan & Brian W. Kernighan.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

//package popcount_test

import (
	"testing"

	"gopl.io/ch2/popcount"
)

// -- Alternative implementations --

func BitCount(x uint64) int {
	// Hacker's Delight, Figure 5-2.
	x = x - ((x >> 1) & 0x5555555555555555)
	x = (x & 0x3333333333333333) + ((x >> 2) & 0x3333333333333333)
	x = (x + (x >> 4)) & 0x0f0f0f0f0f0f0f0f
	x = x + (x >> 8)
	x = x + (x >> 16)
	x = x + (x >> 32)
	return int(x & 0x7f)
}

func PopCountByClearing(x uint64) int {
	n := 0
	for x != 0 {
		x = x & (x - 1) // clear rightmost non-zero bit
		n++
	}
	return n
}

func PopCountByShifting(x uint64) int {
	n := 0
	for i := uint(0); i < 64; i++ {
		if x&(1<<i) != 0 {
			n++
		}
	}
	return n
}

// -- Benchmarks --

func BenchmarkPopCount(b *testing.B) {
	for i := 0; i < b.N; i++ {
		popcount.PopCount(0x1234567890ABCDEF)
	}
}

func BenchmarkBitCount(b *testing.B) {
	for i := 0; i < b.N; i++ {
		BitCount(0x1234567890ABCDEF)
	}
}

func BenchmarkPopCountByClearing(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountByClearing(0x1234567890ABCDEF)
	}
}

func BenchmarkPopCountByShifting(b *testing.B) {
	for i := 0; i < b.N; i++ {
		PopCountByShifting(0x1234567890ABCDEF)
	}
}

// Go 1.6, 2.67GHz Xeon
// $ go test -cpu=4 -bench=. gopl.io/ch2/popcount
// BenchmarkPopCount-4                  200000000         6.30 ns/op
// BenchmarkBitCount-4                  300000000         4.15 ns/op
// BenchmarkPopCountByClearing-4        30000000         45.2 ns/op
// BenchmarkPopCountByShifting-4        10000000        153 ns/op
//
// Go 1.6, 2.5GHz Intel Core i5
// $ go test -cpu=4 -bench=. gopl.io/ch2/popcount
// BenchmarkPopCount-4                  200000000         7.52 ns/op
// BenchmarkBitCount-4                  500000000         3.36 ns/op
// BenchmarkPopCountByClearing-4        50000000         34.3 ns/op
// BenchmarkPopCountByShifting-4        20000000        108 ns/op
//
// Go 1.7, 3.5GHz Xeon
// $ go test -cpu=4 -bench=. gopl.io/ch2/popcount
// BenchmarkPopCount-12                 2000000000        0.28 ns/op
// BenchmarkBitCount-12                 2000000000        0.27 ns/op
// BenchmarkPopCountByClearing-12       100000000        18.5 ns/op
// BenchmarkPopCountByShifting-12       20000000         70.1 ns/op

In [28]:
!go test -cpu=2 -bench=. gopl.io/ch2/popcount

goos: darwin
goarch: arm64
pkg: gopl.io/ch2/popcount
BenchmarkPopCount-2             	1000000000	         0.3207 ns/op
BenchmarkBitCount-2             	1000000000	         0.3165 ns/op
BenchmarkPopCountByClearing-2   	100000000	        11.32 ns/op
BenchmarkPopCountByShifting-2   	55486626	        21.40 ns/op
PASS
ok  	gopl.io/ch2/popcount	3.375s


#### BitCount
* See The Quest for an Accelerated Population Count by Henry S. Warren, Jr. chapter 10 in Beautiful Code.
* [discussion here](https://stackoverflow.com/questions/1458314/number-of-1s-in-32-bit-number)
* [briefing](https://blog.csdn.net/weixin_42092787/article/details/106607426)

<code>
$ go test -cpu=2 -bench=. gopl.io/ch2/popcount
goos: linux
goarch: amd64
pkg: gopl.io/ch2/popcount
BenchmarkPopCount-2                     1000000000               0.704 ns/op
BenchmarkBitCount-2                     1000000000               0.361 ns/op
BenchmarkPopCountByClearing-2           63276454                18.3 ns/op
BenchmarkPopCountByShifting-2           21629844                56.6 ns/op
PASS
ok      gopl.io/ch2/popcount    3.661s
</code>

## 2.7. Scope

In [29]:
import "fmt"
func f() {}
var g = "g"

func main(){
    f := "f"
    fmt.Println(f) // "f"; local var f shadows package-level func f
    fmt.Println(g) // "g"; package-level var
    //fmt.Println(h) // compile error: undefined: h
}

f
g


In [42]:
main()

f
g


In [3]:
// three different variables called x for each declaration appears in a different lexical block
func main(){
    x := "hello world!"
    for i:= 0; i < len(x); i++ {
        x := x[i]
        if x != '!' {
            x := x + 'A' - 'a'
            fmt.Printf("%c", x)   // "HELLO" (one letter per iteration)
        }
    }
    fmt.Printf(" %d", ' ' + 'A' - 'a')
}

HELLOWORLD 0

In [53]:
main()

HELLOWORLD 0

In [4]:
func main(){
    x := "hello-world"
    for _, x := range x { 
        x := x + 'A' - 'a' + 2
        fmt.Printf("%c", x)  // "HELLO" (one letter per iteration)
    }
}

JGNNQYQTNF

In [60]:
main()

JGNNQYQTNF

In [30]:
import "fmt"
func f() int { return 1}
func g(x int) int { return 1}
%%
if x:= f(); x == 0 {
    fmt.Println(x)
} else if y:= g(x); x == y {
    fmt.Println(x,y)
}
fmt.Println(x, y) // compile error: x and y are not visible here

# gonb_0425d04a 

 

 Cell[30]: Line 3 
 ./main.go:39:6: g redeclared in this block
 
	return (f - 32) * 5 / 9
}

 func g(x int) int { return 1}
 
func init(){

 

 

 Cell[29]: Line 3 
 ./main.go:16:2: other declaration of g
 

var (
	cwd string
 g = "g"
 n = flag.Bool("n", false, "omit trailing newline")
	pc [256]byte

 

 

 Cell[30]: Line 7 
 ./main.go:73:16: invalid operation: cannot call non-function g (variable of type string)
 
	flag.Parse()
	if x:= f(); x == 0 {
 fmt.Println(x)
 } else if y:= g(x); x == y {
 fmt.Println(x,y)
	}

 

 

 Cell[30]: Line 10 
 ./main.go:76:14: undefined: x
 
	} else if y:= g(x); x == y {
 fmt.Println(x,y)
	}
 fmt.Println(x, y) // compile error: x and y are not visible here
 
}

 

 

 Cell[30]: Line 10 
 ./main.go:76:17: undefined: y
 
	} else if y:= g(x); x == y {
 fmt.Println(x,y)
	}
 fmt.Println(x, y) // compile error: x and y are not visible here
 
}

ERROR: failed to run "/opt/homebrew/bin/go build -o /var/folders/4d/tg201k8x6yb81f_lrvt4440m0000gn/T/gonb_0425d04a/gonb_0425d04a": exit status 1

In [31]:
//the scope of f is just the if statement, not accessible outside of if
import "os"
%%
    fname := "dup.txt"
    if f, err := os.Open(fname); err != nil {
        //return err
    } // f was never used
    f.Stat()  //compile error: undefined f
    f.Close() //compile error: undefined f

# gonb_0425d04a 

 

 Cell[31]: Line 5 
 ./main.go:70:9: f declared but not used
 
func main() {
	flag.Parse()
 fname := "dup.txt"
 if f, err := os.Open(fname); err != nil {
 //return err
 } // f was never used

 

 

 Cell[31]: Line 8 
 ./main.go:73:8: f.Stat undefined (type func() has no field or method Stat)
 
 if f, err := os.Open(fname); err != nil {
 //return err
 } // f was never used
 f.Stat() //compile error: undefined f
 f.Close() //compile error: undefined f


 

 

 Cell[31]: Line 9 
 ./main.go:74:8: f.Close undefined (type func() has no field or method Close)
 
 //return err
 } // f was never used
 f.Stat() //compile error: undefined f
 f.Close() //compile error: undefined f
 
}

ERROR: failed to run "/opt/homebrew/bin/go build -o /var/folders/4d/tg201k8x6yb81f_lrvt4440m0000gn/T/gonb_0425d04a/gonb_0425d04a": exit status 1

In [34]:
// declare f before the condition, and accessible later
%%
f, err := os.Open("dup.txt")
if err != nil {
    //return err
}
f.Stat()
f.Close()

In [5]:
// avoid declaring f in the outer block, by moving to else block
%%
fname := "dup.txt"
if f, err := os.Open(fname); err != nil {
    //return err
} else {
    // f and err are visible here too
    f.Stat()
    f.Close()
}

In [35]:
import ( "os"; "log")
var cwd string
func init() {
    cwd, err := os.Getwd() // compile error: unused: cwd
    if err != nil {
        log.Fatalf("os.Getwd failed: %v", err)
    }//cwd and err are local in init, package level cwd is not updated
}

# gonb_0425d04a 

 

 Cell[35]: Line 4 
 ./main.go:40:5: cwd declared but not used
 
}

func init() {
 cwd, err := os.Getwd() // compile error: unused: cwd
 if err != nil {
 log.Fatalf("os.Getwd failed: %v", err)

ERROR: failed to run "/opt/homebrew/bin/go build -o /var/folders/4d/tg201k8x6yb81f_lrvt4440m0000gn/T/gonb_0425d04a/gonb_0425d04a": exit status 1

In [1]:
import "os"
import "log"
import "fmt"
var cwd string
func init1() {
//func main() {
    cwd, err := os.Getwd()
    if err != nil {
        //log.Fatalf("os.Getwd failed: %v", err) // log output obfuscated the bug
        //.Fatalf("os.Getwd failed: %v", err) // log output obfuscated the bug
        fmt.Printf("%v",err)
        return

    }
    fmt.Printf("init: Working diectory = %s\n", cwd)
    //log.Printf("Working diectory = %s", cwd)

}
func main(){
    init1()
    fmt.Printf("main: Working diectory = %s\n", cwd)
}
//main()

init: Working diectory = /Users/skhuang/course/go2023/go2023/ch2
main: Working diectory = 


In [2]:
var cwd string    // avoid := by declaring err in a separate var declaration
func init(){
    var err error
    cwd, err = os.Getwd()
    if err != nil {
        log.Fatalf("os.Getwd failed: %v", err)
    }
}